### ISSUE TITLE
Helm values file from external Git repository source and revision missing in Application sourceTypes and revisions  

### ISSUE THREAD

**Issue by RafaelBroseghini** [argoproj/argo-cd#14109]
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

We experience that in multisource apps starting in v2.7.5, when using helm values file from external Git repository, the source revision is missing in the application revisions.

**To Reproduce**

Example multisource:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: guestbook
spec:
  destination:
    namespace: default
    server: https://kubernetes.default.svc
  project: default
  sources:
  - chart: prometheus
    helm:
      valueFiles:
      - $values/values.yaml
    repoURL: https://prometheus-community.github.io/helm-charts
    targetRevision: 15.7.1
  - ref: values
    repoURL: https://github.com/RafaelBroseghini/temp.git
    targetRevision: HEAD
  - chart: elasticsearch
    repoURL: https://helm.elastic.co
    targetRevision: 8.5.1
```

when inspecting the application, we get:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"argoproj.io/v1alpha1","kind":"Application","metadata":{"annotations":{},"name":"guestbook","namespace":"argocd"},"spec":{"destination":{"namespace":"default","server":"https://kubernetes.default.svc"},"project":"default","sources":[{"chart":"prometheus","helm":{"valueFiles":["$values/values.yaml"]},"repoURL":"https://prometheus-community.github.io/helm-charts","targetRevision":"15.7.1"},{"ref":"values","repoURL":"https://github.com/RafaelBroseghini/temp.git","targetRevision":"HEAD"},{"chart":"elasticsearch","repoURL":"https://helm.elastic.co","targetRevision":"8.5.1"}]}}
  creationTimestamp: "2023-06-17T20:40:54Z"
  generation: 3
  name: guestbook
  namespace: argocd
  resourceVersion: "8893"
  uid: 926cff44-1759-4590-b897-6863d23f19b6
spec:
  destination:
    namespace: default
    server: https://kubernetes.default.svc
  project: default
  sources:
  - chart: prometheus
    helm:
      valueFiles:
      - $values/values.yaml
    repoURL: https://prometheus-community.github.io/helm-charts
    targetRevision: 15.7.1
  - ref: values
    repoURL: https://github.com/RafaelBroseghini/temp.git
    targetRevision: HEAD
  - chart: elasticsearch
    repoURL: https://helm.elastic.co
    targetRevision: 8.5.1
status:
  health:
    status: Progressing
  reconciledAt: "2023-06-17T20:40:54Z"
  resources:
 <snip>
  sourceTypes:
  - Helm
  - Helm
  summary:
    images:
    - docker.elastic.co/elasticsearch/elasticsearch:8.5.1
    - jimmidyson/configmap-reload:v0.5.0
    - k8s.gcr.io/kube-state-metrics/kube-state-metrics:v2.3.0
    - prom/pushgateway:v1.4.2
    - quay.io/prometheus/alertmanager:v0.23.0
    - quay.io/prometheus/node-exporter:v1.3.0
    - quay.io/prometheus/prometheus:v2.34.0
  sync:
    comparedTo:
      destination:
        namespace: default
        server: https://kubernetes.default.svc
      sources:
      - chart: prometheus
        helm:
          valueFiles:
          - $values/values.yaml
        repoURL: https://prometheus-community.github.io/helm-charts
        targetRevision: 15.7.1
      - ref: values
        repoURL: https://github.com/RafaelBroseghini/temp.git
        targetRevision: HEAD
      - chart: elasticsearch
        repoURL: https://helm.elastic.co
        targetRevision: 8.5.1
    revisions:
    - 15.7.1
    - 8.5.1
    status: Synced

```

**Expected behavior**

I would expect the application revisions to return:

```yaml
revisions:
  - 15.7.1
  - 627613d1906e57984e0443049e48f5b789852bdf
  - 8.5.1
```

**Screenshots**

**Version**

```shell
Paste the output from `argocd version` here.

argocd: v2.7.5+a2430af
  BuildDate: 2023-06-16T15:00:03Z
  GitCommit: a2430af1c356b283e5e3fc5bde1f5e2b5199f258
  GitTreeState: clean
  GoVersion: go1.19.10
  Compiler: gc
  Platform: linux/amd64
argocd-server: v2.7.5+a2430af
  BuildDate: 2023-06-16T15:00:03Z
  GitCommit: a2430af1c356b283e5e3fc5bde1f5e2b5199f258
  GitTreeState: clean
  GoVersion: go1.19.10
  Compiler: gc
  Platform: linux/amd64
  Kustomize Version: could not get kustomize version: exec: "kustomize": executable file not found in $PATH
  Helm Version: v3.11.2+g912ebc1
  Kubectl Version: v0.24.2
  Jsonnet Version: v0.19.1
```

**Logs**

```
Paste any relevant application logs here.
```

**Comment by RafaelBroseghini** [comment]
There was an attempt to solve this in https://github.com/argoproj/argo-cd/issues/13859, but perhaps it was only fixed for sourceTypes of types: https://github.com/argoproj/argo-cd/blob/8f63ba3ee78384f77e7866c04c27aea08ef29dad/pkg/apis/application/v1alpha1/types.go#L270-L275

Since the source type is missing for external values file, I assume it is not being detected for external values files.

This is not true when using other types of sources in multi source (not external values file), for example:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: guestbook
  namespace: argocd
spec:
  project: default
  destination:
    server: https://kubernetes.default.svc
    namespace: default
  sources:
    - chart: elasticsearch
      repoURL: https://helm.elastic.co
      targetRevision: 8.5.1
    - repoURL: https://github.com/argoproj/argocd-example-apps.git
      path: guestbook
      targetRevision: HEAD
```

When inspecting the application, we get the expected result:

```yaml
  <snip>
    revisions:
    - 8.5.1
    - 53e28ff20cc530b9ada2173fbbd64d48338583ba
```

**Comment by jannfis** [comment]
Related: #14108

**Comment by cjc7373** [comment]
IIUC this is unrelated to external helm values files and will be solved by #14113

**Comment by RafaelBroseghini** [comment]
> IIUC this is unrelated to external helm values files and will be solved by #14113

I believe #14113 is solving retaining the order of revisions, but not the issue that external values file revision is missing in `revisions`.

**Comment by crenshaw-dev** [comment]
@RafaelBroseghini I believe #14113 solves both issues, per this comment: https://github.com/argoproj/argo-cd/pull/14113#issuecomment-1598121774
